File: Napakalaki_Java/NapakalakiGame/BadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.*;
/**
 *
 * @author l
 */
public class BadConsequence {
    

    String text;
    int levels;
    int nVisibleTreasures;
    int nHiddenTreasures;
    ArrayList<TreasureKind> specificVisibleTreasures= new ArrayList();
    ArrayList<TreasureKind> specificHiddenTreasures= new ArrayList();
    boolean death;
    
    
    public BadConsequence(String t, int l, ArrayList<TreasureKind> v,
                ArrayList<TreasureKind> h){
        this.text=t;
        this.levels=l;
        this.specificVisibleTreasures=v;
        this.specificHiddenTreasures=h;
        this.nVisibleTreasures=-1;
        this.nHiddenTreasures=-1;

    }
    
    public BadConsequence(String t, int l, int nVisible, int nHidden){
        this.text=t;
        this.levels=l;
        this.nVisibleTreasures=nVisible;
        this.nHiddenTreasures=nHidden;
        this.specificVisibleTreasures=new ArrayList();
        this.specificHiddenTreasures=new ArrayList();
    }
    
    public BadConsequence(String text, boolean death){
        this.text=text;
        this.death=death;
        this.nVisibleTreasures=-1;
        this.nHiddenTreasures=-1;
        this.specificVisibleTreasures=null;
        this.specificHiddenTreasures=null;        
    }   
    
    
    public boolean isEmpty(){
        if((nVisibleTreasures<1 && nHiddenTreasures<1) && (specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty())){
             return true;
        }
        else return false;
    }  
    
     
    public int getLevels(){
        return levels;
        
    }
    
    public String getText(){
        return text;
    }
    
    public int getNVisibleTreasures(){
        return nVisibleTreasures;
    }
    
    public int getNHiddenTreasures(){
        return nHiddenTreasures;
    }


    public ArrayList<TreasureKind>  getSpecificHiddenTreasures (){
        return specificHiddenTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures(){
        return specificVisibleTreasures;
    }

    public void substractVisibleTreasure( Treasure t){ //contador
        if(nVisibleTreasures<0)
          specificVisibleTreasures.remove(t.getType());
        else nVisibleTreasures--;
        
    }
    
    public void substractHiddenTreasure( Treasure t){
        if(nHiddenTreasures<0)
          specificHiddenTreasures.remove(t.getType());
        else nHiddenTreasures--;
    }
    
    
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h ){
        
        ArrayList<TreasureKind> aux_visible=new ArrayList();
        ArrayList<TreasureKind> aux_hidden=new ArrayList();
        
        if(!specificVisibleTreasures.isEmpty() || !specificHiddenTreasures.isEmpty()){
        for(Treasure treasure : v){
            for(TreasureKind specificTreasure : specificVisibleTreasures)
                if(treasure.getType()==specificTreasure)
                    aux_visible.add(specificTreasure);
        }
        for(Treasure treasure : h){
            for(TreasureKind specificTreasure : specificHiddenTreasures)
                if(treasure.getType()==specificTreasure)
                    aux_hidden.add(specificTreasure);
        }        
        
        }
        else if(nVisibleTreasures>0 || nHiddenTreasures>0){
         
            for(int i=0; i<nVisibleTreasures && i<v.size(); i++)
                aux_visible.add(v.get(i).getType());
            for(int i=0; i<nHiddenTreasures && i<h.size(); i++)
                aux_hidden.add(h.get(i).getType());
        }
        BadConsequence bad=new BadConsequence(" ",0,aux_visible,aux_hidden);
        return bad;
    }

    public boolean myBadConsequenceIsDeath(){
        if (this.death)
            return true;
        else 
            return false;
                   
    }
    
    public String toString(){
        
        List list_hidden=Arrays.asList(specificHiddenTreasures);
        List list_visible=Arrays.asList(specificVisibleTreasures);
        
    return "Text = " + text + " levels = " + Integer.toString(levels) +
            " nVisibleTreasure= " + Integer.toString(nVisibleTreasures) +
            " nHiddenTreasure= " + Integer.toString(nHiddenTreasures) + " death= " +
            Boolean.toString(death) + " specificHiddenTreasures= " + list_hidden
            + " specificVisibleTreasures= " + list_visible ;

           
        }
    
    
}

File: Napakalaki_Ruby/Bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'Card_dealer.rb'
require_relative 'Bad_consequence.rb'
require_relative 'CombatResult.rb'
require_relative 'dice.rb'
require_relative 'monster.rb'
require_relative 'Napakalaki.rb'
require_relative 'player.rb'
require_relative 'prize.rb'
require_relative 'treasure.rb'
require_relative 'treasure_kind.rb'
module NapakalakiGame
  

class BadConsequence
  
  attr_reader :levels, :nVisibleTreasures, :nHiddenTreasures, :specificHiddenTreasures, :specificVisibleTreasures
  private_class_method :new
  

  
  def initialize( aText, someLevels, someVisibleTreasures, someHiddenTreasures, someSpecificVisibleTreasures,
                  someSpecificHiddenTreasures, death)
    @text=aText
    @levels=someLevels
    @nVisibleTreasures=someVisibleTreasures
    @nHiddenTreasures=someHiddenTreasures
    @specificVisibleTreasures=someSpecificVisibleTreasures
    @specificHiddenTreasures=someSpecificHiddenTreasures
    @death=death
  end
  
  def  getLevels
    return @levels
  end
  
  def getNVisibleTreasures
    return @nVisibleTreasures
  end
  
  def getNHiddenTreasures
    return @nHiddenTreasures
  end
  
  def getSpecificHiddenTreasures
    return @specificHiddenTreasures
  end
  
  def getSpecificVisibleTreasures
    return @specificVisibleTreasures
  end
  
  def self.newLevelNumberOfTreasures(aText,someLevels,someVisibleTreasures,someHiddenTreasures)
    new(aText, someLevels, someVisibleTreasures, someHiddenTreasures, Array.new, Array.new, false)
  end

  def self.newLevelSpecificTreasures(aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    new(aText, someLevels, -1, -1, someSpecificVisibleTreasures, someSpecificHiddenTreasures, false) 
  end

  def self.newDeath(aText)
    new(aText, -1, -1, -1, Array.new , Array.new, true)
  end
  
  def isEmpty
    return ((@nVisibleTreasures<1 and @nHiddenTreasures<1) and ( @specificVisibleTreasures.empty? and @specificHiddenTreasures.empty?))  
  end
  
  
  def substractVisibleTreasure (t)
      if(@nVisibleTreasures<0)
        if t.getType!=TreasureKind::ONEHAND        
        @specificVisibleTreasures.delete(t.getType())
        else
        i=0
        while(@specificVisibleTreasures.at(i)!=TreasureKind::ONEHAND and i<@specificVisibleTreasures.size)
        i=i+1
        end
        if(i<@specificVisibleTreasures.size)
          @specificVisibleTreasures.delete_at(i)
        end
        end
      else
      @nVisibleTreasures=@nVisibleTreasures-1
      end
    
  end

  def substractHiddenTreasure (t)
      if(@nHiddenTreasures<0)
        if t.getType!=TreasureKind::ONEHAND
        @specificHiddenTreasures.delete(t.getType())
        else
        i=0
        while(@specificHiddenTreasures.at(i)!=TreasureKind::ONEHAND and i<@specificHiddenTreasures.size)
        i=i+1
        end
        if(i<@specificHiddenTreasures.size)
          @specificHiddenTreasures.delete_at(i)
        end
        end
      else
      @nHiddenTreasures=@nHiddenTreasures-1
      end    
    
  end
  
  def adjustToFitTreasureLists( v, h)        
        aux_visible=Array.new
        aux_hidden=Array.new
        if(!@specificVisibleTreasures.empty? or !@specificHiddenTreasures.empty?)
        v.each{|treasure|
            @specificVisibleTreasures.each{|specificTreasure|
                if(treasure.getType()==specificTreasure)
                    aux_visible << specificTreasure

                end
            }
        }
        h.each{ |treasure|
            @specificHiddenTreasures.each{|specificTreasure|
                if(treasure.getType()==specificTreasure)
                    aux_hidden << specificTreasure
                end
            }
        }
        elsif (@nVisibleTreasures>0 or @nHiddenTreasures>0)
          size=v.size()
          i=0
          while(i<size and i<@nVisibleTreasures)
            aux_visible << v[i].getType()
            i=i+1
          end
          size=h.size()
          i=0
          while(i<size and i<@nHiddenTreasures)
            aux_hidden << h[i].getType()
            i=i+1
          end          
          
        end
        

        
        bad=BadConsequence.newLevelSpecificTreasures("",0,aux_visible,aux_hidden);
        return bad;
  end


    
  
  def myBadConsequenceIsDeath
    if(@death==true)
      return true
    else return false
    end
  end
  
  def to_s
    return "Text= #@text  Levels= #@levels  nVisibleTreasures=#@nVisibleTreasures  nHiddenTreasures=#@nHiddenTreasures  " +
     "death=#@death specificHiddenTreasures=#{@specificHiddenTreasures}  " +
     "specificVisibleTreasures=#{@specificVisibleTreasures}"
  end
  



end


end
